File: api/app/services/events.py
import uuid
from typing import Any

from sqlalchemy.orm import Session

from app.models import BusinessEvent


def append_event(
    db: Session,
    event_type: str,
    payload: dict[str, Any] | None = None,
    org_id: uuid.UUID | None = None,
    actor_user_id: uuid.UUID | None = None,
) -> BusinessEvent:
    ev = BusinessEvent(
        event_type=event_type,
        payload=payload or {},
        org_id=org_id,
        actor_user_id=actor_user_id,
    )
    db.add(ev)
    return ev

File: web/lib/server/events.ts
import type { Prisma } from "@prisma/client";
import { prisma } from "@/lib/prisma";

export async function appendEvent(
  tx: Prisma.TransactionClient,
  eventType: string,
  payload: Prisma.InputJsonValue | null,
  orgId: string | null,
  actorUserId: string | null,
) {
  await tx.businessEvent.create({
    data: {
      eventType,
      payload: payload ?? undefined,
      orgId: orgId ?? undefined,
      actorUserId: actorUserId ?? undefined,
    },
  });
}

/** Outside transactions */
export async function appendEventDirect(
  eventType: string,
  payload: Prisma.InputJsonValue | null,
  orgId: string | null,
  actorUserId: string | null,
) {
  await appendEvent(prisma, eventType, payload, orgId, actorUserId);
}
